src/entities/trades/swapr-v3/SwaprV3.ts

```diff
@@ -0,0 +1,209 @@
+
+import { CurrencyAmount, Percent, Price, TokenAmount } from '../../fractions'
+
+import { TradeWithSwapTransaction } from '../interfaces/trade'
+import { RoutablePlatform } from '../routable-platform'
+import { getProvider, tryGetChainId } from '../utils'
+import { ONE, TradeType } from '../../../constants'
+
+import { UnsignedTransaction } from 'ethers'
+import { TradeOptions } from '../interfaces/trade-options'
+import { parseUnits } from '@ethersproject/units'
+import { Route } from './entities/route'
+import { SWAPR_ALGEBRA_CONTRACTS } from './constants'
+import { getQuoterContract, getRouterContract } from './contracts'
+import { getRoutes } from './routes'
+
+interface SwaprV3ConstructorParams {
+  maximumSlippage: Percent
+  inputAmount: CurrencyAmount
+  outputAmount: CurrencyAmount
+  tradeType: TradeType
+  chainId: number
+  priceImpact: Percent
+  fee: Percent
+}
+
+export class SwaprV3Trade extends TradeWithSwapTransaction {
+  public constructor({
+    inputAmount,
+    outputAmount,
+    maximumSlippage,
+    priceImpact,
+    tradeType,
+    chainId,
+    fee,
+  }: SwaprV3ConstructorParams) {
+    super({
+      details: undefined,
+      type: tradeType,
+      inputAmount,
+      outputAmount,
+      maximumSlippage,
+      platform: RoutablePlatform.SWAPR_V3,
+      chainId,
+      executionPrice: new Price({
+        baseCurrency: inputAmount.currency,
+        quoteCurrency: outputAmount.currency,
+        denominator: inputAmount.raw,
+        numerator: outputAmount.raw,
+      }),
+      priceImpact,
+      fee,
+      approveAddress: SWAPR_ALGEBRA_CONTRACTS['router'],
+    })
+  }
+
+  static async getQuote(
+    { amount, quoteCurrency, tradeType, maximumSlippage }: any,
+    provider?: BaseProvider
+  ): Promise<SwaprV3Trade | null> {
+    const chainId = tryGetChainId(amount, quoteCurrency)
+    invariant(chainId, 'SwaprV3Trade.getQuote: chainId is required')
+
+    maximumSlippage = maximumSlippage || 0
+    provider = provider ?? getProvider(chainId)
+
+    const tokenIn = amount.currency
+    const tokenOut = quoteCurrency
+
+    invariant(
+      (await provider.getNetwork()).chainId == chainId,
+      `SwaprV3Trade.getQuote: currencies chainId does not match provider's chainId`
+    )
+
+    if (tradeType === TradeType.EXACT_INPUT) {
+      const routes: Route<Currency, Currency>[] = await getRoutes(tokenIn, tokenOut, chainId)
+
+      const quotedAmountOut = await getQuoterContract()
+        .callStatic.quoteExactInputSingle(
+          tokenIn.address,
+          tokenOut.address,
+          parseUnits(amount.toSignificant(), amount.currency.decimals),
+          0
+        )
+        .catch((error) => {
+          console.error(`Error sending quoteExactInputSingle transaction: ${error}`)
+          return null
+        })
+
+      const fee =
+        routes?.length > 0 && routes[0].pools.length > 0
+          ? new Percent(routes[0].pools[0].fee.toString(), '1000000')
+          : new Percent('0', '0')
```

Comment:
We can declare this outside the conditions. Both trades types have the same code.